patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs

```diff
@@ -631,4 +637,123 @@ internal static void EraseWorld(string path, bool cloudSave)
+			return;
+
+		try {
+			// this code hard-traverses the NBT format to read the just the first nested tag, if preset.
+			// It relies on deterministic tag saving order for the header to be first.
+			// Because the NBT format is not seekable, there is no way to skip reading the entire tag tree in order to find a specific sub-tag at an arbitrary path.
+
+			using Stream stream = isCloudSave ? new MemoryStream(SocialAPI.Cloud.Read(path)) : new FileStream(path, FileMode.Open);
+			using BinaryReader reader = new BigEndianReader(new GZipStream(stream, CompressionMode.Decompress));
+			if (reader.ReadByte() != 10)
+				throw new IOException("Root tag not a TagCompound");
+
+			// ignore root tag name
+			_ = TagIO.ReadTagImpl(8, reader);
+			if (reader.ReadByte() != 10)
+				return; // no header tag
+
+			if ((string)TagIO.ReadTagImpl(8, reader) != "0header")
+				return;
+
+			LoadWorldHeader(data, (TagCompound)TagIO.ReadTagImpl(10, reader));
+		}
+		catch (Exception ex) {
+			Logging.tML.Warn($"Error reading .twld header from: {path} (IsCloudSave={isCloudSave})", ex);
+		}
+	}
+
+	private static void LoadWorldHeader(WorldFileData data, TagCompound tag)
+	{
+		data.ModHeaders = new Dictionary<string, TagCompound>();
+		foreach (var entry in tag.GetCompound("modHeaders")) {
+			string fullname = entry.Key;
+
+			if (ModContent.TryFind<ModSystem>(fullname, out var system)) // handle legacy renames
+				fullname = system.FullName;
+
+			data.ModHeaders[fullname] = (TagCompound)entry.Value;
+		}
+		LoadUsedMods(data, tag.GetList<string>("usedMods"));
+		LoadUsedModPack(data, tag.GetString("usedModPack"));
+		if(tag.ContainsKey("generatedWithMods")) // GetList will return an empty List instead of null. null and empty list have different meaning for this data.
+			LoadGeneratedWithMods(data, tag.GetList<TagCompound>("generatedWithMods"));
+	}
+
+	internal static void LoadUsedMods(WorldFileData data, IList<string> usedMods)
+	{
+		data.usedMods = usedMods;
+	}
+
+	internal static List<string> SaveUsedMods()
+	{
+		return ModLoader.Mods.Select(m => m.Name).Except(new[] { "ModLoader" }).ToList();
+	}
+
+	internal static void LoadUsedModPack(WorldFileData data, string modpack)
+	{
+		data.modPack = string.IsNullOrEmpty(modpack) ? null : modpack; // tag.GetString returns "" even though null was saved.
+	}
+
+	internal static string SaveUsedModPack()
+	{
+		return Path.GetFileNameWithoutExtension(Core.ModOrganizer.ModPackActive);
+	}
+
+	internal static void LoadGeneratedWithMods(WorldFileData data, IList<TagCompound> list)
+	{
+		data.modsGeneratedWith = list.Select(x => (x.GetString("mod"), new Version(x.GetString("version")))).ToList();
+	}
+
+	internal static List<TagCompound> SaveGeneratedWithMods()
+	{
+		if (Main.ActiveWorldFileData.modsGeneratedWith == null)
+			return null;
+		return Main.ActiveWorldFileData.modsGeneratedWith
+			.Select(x => new TagCompound {
+				["mod"] = x.modName,
+				["version"] = x.modVersion.ToString()
```

Comment:
As mentioned above, in the 'use dictionary' feedback. Consider adding a tag serializer for `Version` if we don't have one instead of manually doing to/from string